File: gto-support-recyclerview/src/main/java/org/ccci/gto/android/common/recyclerview/adapter/SimpleDataBindingAdapter.java
package org.ccci.gto.android.common.recyclerview.adapter;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;

public abstract class SimpleDataBindingAdapter<B extends ViewDataBinding>
        extends RecyclerView.Adapter<DataBindingViewHolder<B>> {
    // region Lifecycle Events

    @NonNull
    @Override
    public final DataBindingViewHolder<B> onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {
        return new DataBindingViewHolder<>(onCreateViewDataBinding(parent, viewType));
    }

    @NonNull
    protected abstract B onCreateViewDataBinding(@NonNull ViewGroup parent, int viewType);

    @Override
    public final void onBindViewHolder(@NonNull final DataBindingViewHolder<B> holder, final int position) {
        onBindViewDataBinding(holder.binding, position);
    }

    protected abstract void onBindViewDataBinding(@NonNull B binding, int position);

    @Override
    public final void onViewRecycled(@NonNull final DataBindingViewHolder<B> holder) {
        onViewDataBindingRecycled(holder.binding);
    }

    protected void onViewDataBindingRecycled(@NonNull final B binding) { }

    // end Lifecycle Events
}

File: gto-support-recyclerview/src/main/kotlin/org/ccci/gto/android/common/recyclerview/adapter/SimpleDataBindingAdapter.kt
package org.ccci.gto.android.common.recyclerview.adapter

import androidx.databinding.ViewDataBinding
import androidx.lifecycle.LifecycleOwner
import org.ccci.gto.android.common.androidx.recyclerview.adapter.DataBindingViewHolder

abstract class SimpleDataBindingAdapter<B : ViewDataBinding>(lifecycleOwner: LifecycleOwner? = null) :
    AbstractDataBindingAdapter<B, DataBindingViewHolder<B>>(lifecycleOwner) {
    final override fun onCreateViewHolder(binding: B, viewType: Int) = DataBindingViewHolder(binding)
    final override fun onBindViewHolder(holder: DataBindingViewHolder<B>, binding: B, position: Int) =
        onBindViewDataBinding(binding, position)
    final override fun onViewRecycled(holder: DataBindingViewHolder<B>, binding: B) = onViewDataBindingRecycled(binding)

    protected abstract fun onBindViewDataBinding(binding: B, position: Int)
    protected open fun onViewDataBindingRecycled(binding: B) = Unit
}
